### DISCUSSION TITLE
cpp/basics/sizeof

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#23] [Announcements]
# cpp/basics/sizeof

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/sizeof.html

**Comment by Xenorian** [comment]
sizeof那里，如果有虚函数的话，会多一个虚函数表指针

**Reply by imarvinle** [reply]
👍是的，但是大多数考这种主要是考察 C语言 struct，不过这里可以补充一下

**Comment by bostonrayAlen** [comment]
sizeof 感觉不准确补充下。
AMD64 cpu 是64位的,64位虚拟地址. 可以在上面运行64位 Windowst 、32位 Windowst、 DOS（16bit）。取决于操作系统也是不准确的 ：64位 Windows 可以支持 32位应用程序 和 16位 DOS 程序 。取决于编译器基本正确。但是一种编译器可以为多个目标平台生成代码。
准确说应该是与编译器的目标平台有关。如果目标平台是32位的，那么sizeof(void*)就是4，如果是64位的，那么sizeof就是8，如果是16位的，就是2。

**Reply by imarvinle** [reply]
很细👍

**Reply by Aurorxa** [reply]
其实是机器的字长，所谓的 8 位、16位、32位、64 位都是指的是机器的字长。

**Comment by LittleFattyQing** [comment]
64位下：
1. 24
2. 40
3. 1

**Reply by yukifaye** [reply]
应该是17 第一个

**Reply by Nizhenjie** [reply]
第一个是24，大小得是所有类型的最小公倍数

**Reply by huangjiawei128** [reply]
第一个是24的解释：为了保证double b与8字节对齐，int a会pad到8字节，这样整体是8 + 8 + 1 = 17字节，又为了保证整个结构体与最大的成员变量double b的自然对齐边界对齐，整体会pad到24字节
第三个是1的理由：在 C++ 中，当你定义一个空结构体（即不包含任何成员变量的结构体）时，sizeof(EmptyStruct) 返回 1。这是因为 C++ 标准要求每个不同的对象必须占用不同的地址，即使是空结构体。因此，编译器会为空结构体分配一个字节的内存，以确保每个实例都有一个唯一的地址。
